dali/pipeline/operators/geometric/shapes.h

```diff
@@ -0,0 +1,134 @@
+#include "dali/pipeline/operators/operator.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+template <typename Backend>
+class Shapes : public Operator<Backend> {
+ public:
+  Shapes(const Shapes &) = delete;
+  explicit Shapes(const OpSpec &spec) : Operator<Backend>(spec) {
+    output_type_ = spec.GetArgument<DALIDataType>("type");
+    switch (output_type_) {
+    case DALI_INT32:
+    case DALI_UINT32:
+    case DALI_INT64:
+    case DALI_UINT64:
+    case DALI_FLOAT:
+    case DALI_FLOAT64:
+      break;
+    default:
+      {
+        auto &name = TypeTable::GetTypeInfo(output_type_).name();
+        DALI_FAIL("Operator Shapes can return the output as one of the following:\n"
+          "int32, uint32, int64, uint64, float or double;\n"
+          "requested: " + name);
+        break;
+      }
+    }
+  }
+  bool CanInferOutputs() const override { return true; }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+    output_desc[0].type = TypeTable::GetTypeInfo(output_type_);
+    decltype(auto) shape = GetInputShape(ws);
+    output_desc[0].shape = ShapeShape(shape);;
+    return true;
+  }
+
+  void RunImpl(workspace_t<Backend> &ws) override {
+    RunBackend(ws);
+  }
+
+  template <typename type>
+  void ConvertShape(TensorList<CPUBackend> &out, const kernels::TensorListShape<> &shape) {
+    assert(out.shape().num_samples() == shape.num_samples());
+    for (int i = 0; i < shape.num_samples(); i++) {
+      type *data = out.mutable_tensor<type>(i);
+      auto tshape = shape.tensor_shape_span(i);
+      for (int j = 0; j < shape.sample_dim(); j++)
+        data[j] = tshape[j];
+    }
+  }
+
+  void ConvertShape(TensorList<CPUBackend> &out, const kernels::TensorListShape<> &shape) {
+    TYPE_SWITCH(output_type_, type2id, type,
+                (int32_t, uint32_t, int64_t, uint64_t, float, double),
+      (ConvertShape<type>(out, shape);),
+      (DALI_FAIL("Unsupported type for Shapes")));
+  }
+
+  void ConvertShapes(const workspace_t<Backend> &ws, bool pinned) {
+    if (!tmp_.raw_data()) {
+      auto &type = TypeTable::GetTypeInfo(output_type_);
+      tmp_.set_type(type);
+      tmp_.set_pinned(pinned);
+    }
+
+    auto &output = ws.template OutputRef<Backend>(0);
+    tmp_.Resize(output.shape());
+    ConvertShape(tmp_, GetInputShape(ws));
+  }
+
+  void RunBackend(DeviceWorkspace &ws) {
+    ConvertShapes(ws, true);
+    auto &output = ws.template OutputRef<Backend>(0);
+    output.Copy(tmp_, ws.stream());
+  }
+
+  void RunBackend(HostWorkspace &ws) {
+    ConvertShapes(ws, false);
+    auto &output = ws.template OutputRef<Backend>(0);
+    for (int i = 0 ; static_cast<size_t>(i) < tmp_.ntensor(); i++) {
+      auto &output_tensor = output[i];
+      auto *dest = output_tensor.raw_mutable_data();
+      auto *src = tmp_.raw_mutable_tensor(i);
+      std::memcpy(dest, src, output_tensor.nbytes());
```

Comment:
Why do the `batch_size` memcpys instead of writing directly to output in case of CPU op?

Reply:
Because of TensorVector. I can rework ConvertShape to work with it.